addons/l10n_dk_oioubl/tests/test_xml_oioubl_dk.py

```diff
@@ -0,0 +1,168 @@
+from freezegun import freeze_time
+
+from odoo import Command
+from odoo.addons.l10n_account_edi_ubl_cii_tests.tests.common import TestUBLCommon
+from odoo.addons.account.tests.test_account_move_send import TestAccountMoveSendCommon
+from odoo.exceptions import UserError
+from odoo.tests import tagged
+
+
+@tagged('post_install_l10n', 'post_install', '-at_install')
+class TestUBLDK(TestUBLCommon, TestAccountMoveSendCommon):
+
+    @classmethod
+    def setUpClass(cls, chart_template_ref="dk"):
+        super().setUpClass(chart_template_ref=chart_template_ref)
+        cls.company_data['company'].write({
+            'country_id': cls.env.ref('base.dk').id,
+            'currency_id': cls.env.ref('base.DKK').id,
+            'city': 'Aalborg',
+            'zip': '9430',
+            'vat': 'DK12345674',
+            'phone': '+45 32 12 34 56',
+            'street': 'Paradisæblevej, 10',
+            'invoice_is_ubl_cii': True,
+        })
+        cls.env['res.partner.bank'].create({
+            'acc_type': 'iban',
+            'partner_id': cls.company_data['company'].partner_id.id,
+            'acc_number': 'DK5000400440116243',
+        })
+
+        cls.partner_a.write({
+            'name': 'SUPER DANISH PARTNER',
+            'city': 'Aalborg',
+            'zip': '9430',
+            'vat': 'DK12345674',
+            'phone': '+45 32 12 35 56',
+            'street': 'Paradisæblevej, 11',
+            'country_id': cls.env.ref('base.dk').id,
+            'ubl_cii_format': 'oioubl_201',
+        })
+        cls.partner_b.write({
+            'name': 'SUPER BELGIAN PARTNER',
+            'street': 'Rue du Paradis, 10',
+            'zip': '6870',
+            'city': 'Eghezee',
+            'country_id': cls.env.ref('base.be').id,
+            'phone': '061928374',
+            'vat': 'BE0897223670',
+            'ubl_cii_format': 'oioubl_201',
+        })
+        cls.partner_c = cls.env["res.partner"].create({
+            'name': 'SUPER FRENCH PARTNER',
+            'street': 'Rue Fabricy, 16',
+            'zip': '59000',
+            'city': 'Lille',
+            'country_id': cls.env.ref('base.fr').id,
+            'phone': '+33 1 23 45 67 89',
+            'vat': 'FR23334175221',
+            'company_registry': '123 568 941 00056',
+            'ubl_cii_format': 'oioubl_201',
+        })
+        cls.dk_local_tax_1 = cls.env["account.chart.template"].ref('tax120')
+        cls.dk_local_tax_2 = cls.env["account.chart.template"].ref('tax110')
+        cls.dk_foreign_tax_1 = cls.env["account.chart.template"].ref('tax210')
+        cls.dk_foreign_tax_2 = cls.env["account.chart.template"].ref('tax220')
+
+    def create_post_and_send_invoice(self, partner=None, move_type='out_invoice'):
```

Comment:
Any chance to reuse `_generate_move` or make it more useful? I don't really like to have a custom method creating an invoice in every UBL :/

Reply:
Just copy-pasting what we said on discord for the records:

There is a lot of stuff I don't like in `_generate_move` and that would make the tests harder to read.
- naming: it's called `_generate_move` but it does a lot more than that
- api: `seller`, `buyer` is confusing
- api: `invoice_line_ids` have to be passed which makes each test a bit more heavy to read and understand
- `_generate_move` is writing the `seller` info on the company each time it's called
- To test `UserError`, I have to set `from_cron=False` and `allow_fallback_pdf=False` when calling `_generate_pdf_and_send_invoice`

I could refactor `_generate_move` but I will probably run into further more trouble that I will have to solve and that would delay the pr even more and we can't afford that at the moment.